tests/trainer/test_trainer.py

```diff
@@ -518,6 +521,106 @@ def on_load_checkpoint(self, _):
+    trainer = Trainer(**trainer_args)
+    trainer.fit(model, datamodule=dm)
+
+    resume_ckpt = str(tmpdir / "last.ckpt")
+    state_dict = torch.load(resume_ckpt)
+
+    trainer_args.update(
+        {"max_epochs": 3, "resume_from_checkpoint": resume_ckpt, "checkpoint_callback": False, "callbacks": []}
+    )
+
+    class CustomClassifModel(CustomClassifModel):
+        def _is_equal(self, a, b):
+            if isinstance(a, torch.Tensor):
+                return torch.all(torch.eq(a, b))
+
+            if isinstance(a, Mapping):
+                return all(self._is_equal(a.get(k, None), b.get(k, None)) for k in b.keys())
+
+            return a == b
+
+        def _check_optimizers(self):
+            return all(
+                self._is_equal(self.trainer.optimizers[i].state_dict(), state_dict["optimizer_states"][i])
+                for i in range(len(self.trainer.optimizers))
+            )
+
+        def _check_schedulers(self):
+            return all(
+                self._is_equal(self.trainer.lr_schedulers[i]["scheduler"].state_dict(), state_dict["lr_schedulers"][i])
+                for i in range(len(self.trainer.lr_schedulers))
+            )
+
+        def _check_model_state_dict(self):
+            for k in self.state_dict().keys():
+                yield self._is_equal(self.state_dict()[k], state_dict["state_dict"][k])
+
+        def _test_on_val_test_predict_tune_start(self):
+            assert self.trainer.current_epoch == state_dict["epoch"]
+            assert self.trainer.global_step == state_dict["global_step"]
+            assert not any(self._check_model_state_dict())
+
+            # no optimizes and schedulers are loaded otherwise
+            if self.trainer.state.fn != TrainerFn.TUNING:
+                return
+
+            assert not self._check_optimizers()
+            assert not self._check_schedulers()
+
+        def on_train_start(self):
+            if self.trainer.state.fn == TrainerFn.TUNING:
+                self._test_on_val_test_predict_tune_start()
+            else:
+                assert self.trainer.current_epoch == state_dict["epoch"]
+                assert self.trainer.global_step == state_dict["global_step"]
+                assert all(self._check_model_state_dict())
+                assert self._check_optimizers()
+                assert self._check_schedulers()
+
+        def on_validation_start(self):
+            if self.trainer.state.fn == TrainerFn.VALIDATING:
+                self._test_on_val_test_predict_tune_start()
+
+        def on_test_start(self):
+            self._test_on_val_test_predict_tune_start()
+
+    for fn in ("tune", "fit", "validate", "test", "predict"):
+        model = CustomClassifModel()
+        dm = ClassifDataModule()
+        trainer_args.update(
+            {
+                "auto_scale_batch_size": True if fn == "tune" else False,
+            }
+        )
+
+        trainer = Trainer(**trainer_args)
+        getattr(trainer, fn)(model, datamodule=dm)
```

Comment:
nit: split this just in case one fails

```suggestion
        trainer_fn = getattr(trainer, fn)
        trainer_fn(model, datamodule=dm)
```